pdal/PyMesh.cpp

```diff
@@ -0,0 +1,216 @@
+* OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+* AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+* OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+* OF SUCH DAMAGE.
+****************************************************************************/
+
+#include "PyMesh.hpp"
+
+#include <numpy/arrayobject.h>
+
+
+namespace pdal
+{
+namespace python
+{
+
+// Create new empty mesh
+//
+Mesh::Mesh() : m_mesh(nullptr)
+{
+    hasMesh = false;
+    if (_import_array() < 0)
+        throw pdal_error("Could not import numpy.core.multiarray.");
+}
+
+Mesh::~Mesh()
+{
+    if (m_mesh)
+        Py_XDECREF((PyObject *)m_mesh);
+}
+
+// Update from a PointView
+//
+void Mesh::update(PointViewPtr view)
+{
+    npy_intp size;
+
+    if (m_mesh)
+        Py_XDECREF((PyObject *)m_mesh);
+    m_mesh = nullptr;  // Just in case of an exception.
+
+    TriangularMesh* mesh = view->mesh();
+    if (mesh)
+    {
+        hasMesh = true;
+        size = mesh->size();
+    } else {
+        hasMesh = false;
+        size = 0;
+    }
+
+    PyObject *dtype_dict = (PyObject*)buildNumpyDescription(view);
+    if (!dtype_dict)
+        throw pdal_error("Unable to build numpy dtype "
+                "description dictionary");
+
+    PyArray_Descr *dtype = nullptr;
+    if (PyArray_DescrConverter(dtype_dict, &dtype) == NPY_FAIL)
+        throw pdal_error("Unable to build numpy dtype");
+    Py_XDECREF(dtype_dict);
+
+    // This is a 1 x size array.
+    m_mesh = (PyArrayObject *)PyArray_NewFromDescr(&PyArray_Type, dtype,
+            1, &size, 0, nullptr, NPY_ARRAY_CARRAY, nullptr);
+
+    for (PointId idx = 0; idx < size; idx++)
+    {
+        char* p = (char *)PyArray_GETPTR1(m_mesh, idx);
+        const Triangle& t = (*mesh)[idx];
+        uint32_t a = (uint32_t)t.m_a;
+        std::memcpy(p, &a, 4);
+        uint32_t b = (uint32_t)t.m_b;
+        std::memcpy(p + 4, &b, 4);
+        uint32_t c = (uint32_t)t.m_c;
+        std::memcpy(p + 8, &c,  4);
+    }
+}
+
+
+
+PyArrayObject *Mesh::getPythonArray() const
+{
+    return m_mesh;
+}
+
+PyObject* Mesh::buildNumpyDescription(PointViewPtr view) const
+{
+    // Build up a numpy dtype dictionary
+    //
+    // {'formats': ['f8', 'f8', 'f8', 'u2', 'u1', 'u1', 'u1', 'u1', 'u1',
+    //              'f4', 'u1', 'u2', 'f8', 'u2', 'u2', 'u2'],
+    // 'names': ['X', 'Y', 'Z', 'Intensity', 'ReturnNumber',
+    //           'NumberOfReturns', 'ScanDirectionFlag', 'EdgeOfFlightLine',
+    //           'Classification', 'ScanAngleRank', 'UserData',
+    //           'PointSourceId', 'GpsTime', 'Red', 'Green', 'Blue']}
+    //
+
+    Dimension::IdList dims = view->dims();
+
+    PyObject* dict = PyDict_New();
+    PyObject* formats = PyList_New(3);
+    PyObject* titles = PyList_New(3);
+
+    PyList_SetItem(titles, 0, PyUnicode_FromString("A"));
```

Comment:
is `ABC` the common naming for vertices of meshes in Python? Would someone ever want to change this or provide their own?

Reply:
A,B,C is a common convention across all languages and has the advantage of being in line with the PDAL internal data structure. Most Python packages I have seen actually just refer to a tuple of vertices without names, but that would have been less optimal for this API.

I don't think anyone will object or want to change it in the future.